8331087: Move immutable nmethod data from CodeCache

vnkozlov opened the pull request:
Move immutable nmethod's data from CodeCache to C heap. It includes `dependencies, nul_chk_table, handler_table, scopes_pcs, scopes_data, speculations`. It amounts for about 30% (optimized VM) of space in CodeCache.

Use HotSpot's `os::malloc()` to allocate memory in C heap for immutable nmethod's data. Call `vm_exit_out_of_memory()` if allocation failed.

Shuffle fields order and change some fields size from 4 to 2 bytes to avoid nmethod's header size increase.

Tested tier1-5, stress,xcomp

Our performance testing does not show difference.

Example of updated `-XX:+PrintNMethodStatistics` output is in JBS comment.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8331087](https://bugs.openjdk.org/browse/JDK-8331087): Move immutable nmethod data from CodeCache (**Enhancement** - P4)


### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**)
 * [Doug Simon](https://openjdk.org/census#dnsimon) (@dougxc - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18984/head:pull/18984` \
`$ git checkout pull/18984`

Update a local copy of the PR: \
`$ git checkout pull/18984` \
`$ git pull https://git.openjdk.org/jdk.git pull/18984/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18984`

View PR using the GUI difftool: \
`$ git pr show -t 18984`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18984.diff">https://git.openjdk.org/jdk/pull/18984.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18984#issuecomment-2080133031)

vnkozlov commented on test/hotspot/jtreg/compiler/c1/TestLinearScanOrderMain.java:
```diff
@@ -26,6 +26,7 @@
  * @bug 8207355
  * @compile TestLinearScanOrder.jasm
  * @run main/othervm -Xcomp -XX:+TieredCompilation -XX:TieredStopAtLevel=1
+ *                   -XX:+IgnoreUnrecognizedVMOptions -XX:NMethodSizeLimit=655360
```
This test caught one `check_cast<>` issue during development but only on aarch64.
On x64 the test bailed out compilation before that because default `NMethodSizeLimit` was not big enough ((64*K)*wordSize = 524288).

vnkozlov commented on src/hotspot/share/code/nmethod.hpp:
```diff
@@ -524,48 +538,52 @@ class nmethod : public CodeBlob {
-  int handler_table_size() const { return int(          handler_table_end() -           handler_table_begin()); }
-  int nul_chk_table_size() const { return int(          nul_chk_table_end() -           nul_chk_table_begin()); }
+  int immutable_data_size() const { return _immutable_data_size; }
+  int consts_size        () const { return int(          consts_end       () -           consts_begin       ()); }
+  int insts_size         () const { return int(          insts_end        () -           insts_begin        ()); }
+  int stub_size          () const { return int(          stub_end         () -           stub_begin         ()); }
+  int oops_size          () const { return int((address) oops_end         () - (address) oops_begin         ()); }
+  int metadata_size      () const { return int((address) metadata_end     () - (address) metadata_begin     ()); }
+  int scopes_data_size   () const { return int(          scopes_data_end  () -           scopes_data_begin  ()); }
+  int scopes_pcs_size    () const { return int((intptr_t)scopes_pcs_end   () - (intptr_t)scopes_pcs_begin   ()); }
+  int dependencies_size  () const { return int(          dependencies_end () -           dependencies_begin ()); }
+  int handler_table_size () const { return int(          handler_table_end() -           handler_table_begin()); }
+  int nul_chk_table_size () const { return int(          nul_chk_table_end() -           nul_chk_table_begin()); }
```
Shift by one space to aline code.

vnkozlov commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -110,6 +110,11 @@
 
 #endif
 
+// Cast from int value to narrow type
+#define CHECKED_CAST(result, T, thing)      \
+  result = static_cast<T>(thing); \
+  assert(static_cast<int>(result) == thing, "failed: %d != %d", static_cast<int>(result), thing);
+
```
I replaced `checked_cast<>()` with this macro because of next issues:
 - The existing assert points to `utilities/checkedCast.hpp` file where this method is located and not where failed cast. It does not help when it is used several times in one method (for example, in `nmethod()` constructors).
 - The existing assert does not print values

dean-long replied:
I thought @kimbarrett had a draft PR to address the error reporting issue, but I can't seem to find it.  To solve the general problem, I think we need a version of vmassert() that takes `char* file, int lineno` as arguments, and a macro wrapper for checked_cast() that passes `__FILE__` and `__LINEN__` from the caller.

vnkozlov replied:
Yes, it would be perfect separate RFE.

vnkozlov commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -1262,20 +1316,21 @@ nmethod::nmethod(
+    assert((data_offset() + data_end_offset) <= nmethod_size, "wrong nmethod's size: %d < %d", nmethod_size, (data_offset() + data_end_offset));
+
+    // native wrapper does not have read-only data but we need unique not null address
+    _immutable_data          = data_end();
```
I can't use nullptr because VM expects not null address when it checks, for example, `dependencies_begin()` even so sizes are 0.
I used `data_end()` instead of nullptr in other places too.

dean-long commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -1262,20 +1316,21 @@ nmethod::nmethod(
+    _handler_table_offset    = _nul_chk_table_offset;
+    _scopes_pcs_offset       = _handler_table_offset;
     _scopes_data_offset      = _scopes_pcs_offset;
-    _handler_table_offset    = _scopes_data_offset;
-    _nul_chk_table_offset    = _handler_table_offset;
 #if INCLUDE_JVMCI
-    _speculations_offset     = _nul_chk_table_offset;
+    _speculations_offset     = _scopes_data_offset;
     _jvmci_data_offset       = _speculations_offset;
```
Why not use 0 for all these?

vnkozlov replied:
Right. Before all these offsets were assigned to _dependencies_offset which was not 0.
But now we can use 0 for all of them since "_dependencies_offset" is 0.

dean-long commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -1421,25 +1478,40 @@ nmethod::nmethod(
+      // C1 generates UnwindHandler at the end of instructions section.
+      // Calculate positive offset as distance between the start of stubs section
+      // (which is also the end of instructions section) and the start of the handler.
+      CHECKED_CAST(_unwind_handler_offset, int16_t, (_stub_offset - code_offset() - offsets->value(CodeOffsets::UnwindHandler)));
```
```suggestion
      int unwind_handler_offset = code_offset() + offsets->value(CodeOffsets::UnwindHandler);
      CHECKED_CAST(_unwind_handler_offset, int16_t, _stub_offset - unwind_handler_offset);
```

vnkozlov replied:
Will do.

dean-long commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/NMethod.java:
```diff
@@ -516,7 +525,7 @@ public void dumpReplayData(PrintStream out) {
   private int getScopesDataOffset()   { return (int) scopesDataOffsetField  .getValue(addr); }
   private int getScopesPCsOffset()    { return (int) scopesPCsOffsetField   .getValue(addr); }
-  private int getDependenciesOffset() { return (int) dependenciesOffsetField.getValue(addr); }
+  private int getDependenciesOffset() { return (int) 0; }
```
```suggestion
```
No longer used.

vnkozlov replied:
Will remove.

vnkozlov commented:
Thank you, @dean-long, for review. I will collect (hopefully) more comments for next update before testing and pushing it.

dougxc commented on src/hotspot/share/code/nmethod.hpp:
```diff
@@ -463,6 +470,12 @@ class nmethod : public CodeBlob {
+  enum ResultStatus {
+    passed,
+    code_cache_full,
+    out_of_memory
```
Maybe `out_of_malloc_memory` would be clearer? Or `out_of_immutable_data_memory` if immutable data may not always be malloc'ed.

vnkozlov replied:
May be `no_space_for_mmutable_data`.

dougxc commented on src/hotspot/share/jvmci/jvmciRuntime.cpp:
```diff
@@ -2175,6 +2175,7 @@ JVMCI::CodeInstallResult JVMCIRuntime::register_method(JVMCIEnv* JVMCIENV,
                                                         nmethod_entry_patch_offset,
                                                         nmethod_mirror_name,
                                                         failed_speculations);
+      nmethod::ResultStatus result_status;
```
Please expose the new `out_of_memory` result to JVMCI (e.g. in `JVMCI::CodeInstallResult` enum and `HotSpotVMConfig.getCodeInstallResultDescription` method).

vnkozlov commented:
@dean-long and @dougxc  

I am thinking may be I should not bailout  when `malloc` (or other space reservation in a future) failed to allocate memory for immutable data. But instead increase nmethod size and put immutable data there (as before).  Then we bailout only when CodeCache is full as before and we don't need `out_of_memory` failure reason. May be only record that in logs (when they are enabled).

What do you think?

dean-long commented:
It only makes sense if the immutable data heap is not also used for other critical resources.  If malloc or metaspace were used as the immutable data heap, normally failures in those heaps are fatal, because other critical resources (monitors, classes, etc) are allocated from there, so any failure means the JVM is about to die.  There's no reason to find a fall-back method to allocate a new nmethod in that case.

dougxc commented:
Just to be clear @dean-long , you're saying failure to allocate immutable data in the C heap should result in a fatal error? Makes sense to me as the VM must indeed be very close to crashing anyway in that case. It also, obviates the need for propagating `out_of_memory_error` to JVMCI code.

vnkozlov commented:
Update:
1. Addressed @dean-long first comments.
2. Based on discussion with Doug and Tom (see comments in [JDK-8331087](https://bugs.openjdk.org/browse/JDK-8331087)), moved `jvmci_data` back to nmethod's mutable data section.
3. Replaced my allocation failure handling code with call to `vm_exit_out_of_memory()`.

    I verified (with `UseNewCode` hack`) that out of memory is correctly reported in fastdebug and product VMs:
```
#
# There is insufficient memory for the Java Runtime Environment to continue.
# Native memory allocation (malloc) failed to allocate 64 bytes. Error detail: nmethod: no space for immutable data
# An error report file with more information is saved as:
# /scratch/kvn/jdk_git/hs_err_pid4086275.log
```

dean-long commented:
> Just to be clear @dean-long , you're saying failure to allocate immutable data in the C heap should result in a fatal error? Makes sense to me as the VM must indeed be very close to crashing anyway in that case. It also, obviates the need for propagating `out_of_memory_error` to JVMCI code.

I hadn't thought it through that far, actually.  I was only pointing out that the proposed fall-back:

> increase nmethod size and put immutable data there (as before).

isn't worth the trouble.  But making the C heap failure fatal immediately is reasonable, especially if it simplifies JVMCI error reporting.

vnkozlov commented:
> Looks good to me. Did you measure any impact on performance (potentially due to improved code density)?

Thank you, @TobiHartmann, for review.

> What's left for [JDK-7072317](https://bugs.openjdk.org/browse/JDK-7072317)?

Make Relocation info (10% of nmethod size) immutable by moving all encoded pointers (external words and others, which we need to patch in Leyden when loading cached code) from it into separate mutable section.
 
> I wonder if the `CHECKED_CAST` changes shouldn't go into a separate RFE.

No, I want clear information which cast failed instead of trying to reproduce very intermittent failure like this: [JDK-8331253](https://bugs.openjdk.org/browse/JDK-8331253). When you have several `checked_cast<T>` in one method it is impossible to find which failed without this macro.

vnkozlov commented:
Thank you, Dean, Doug and Tobias for reviews.

vnkozlov commented:
/integrate
